openmc/model/triso.py

```diff
@@ -0,0 +1,1335 @@
+        (~0.38).
+    contraction_rate : float, optional
+        Contraction rate of the outer diameter. Higher packing fractions can be
+        reached using a smaller contraction rate, but the algorithm will take
+        longer to converge.
+    seed : int, optional
+        RNG seed.
+
+    Returns
+    ------
+    numpy.ndarray
+        Cartesian coordinates of sphere centers.
+
+    Notes
+    -----
+    The sphere configuration is generated using a combination of random
+    sequential packing (RSP) and close random packing (CRP). RSP performs
+    better than CRP for lower packing fractions (pf), but it becomes
+    prohibitively slow as it approaches its packing limit (~0.38). CRP can
+    achieve higher pf of up to ~0.64 and scales better with increasing pf.
+
+    If the desired pf is below some threshold for which RSP will be faster than
+    CRP ('initial_packing_fraction'), only RSP is used. If a higher pf is
+    required, spheres with a radius smaller than the desired final radius
+    (and therefore with a smaller pf) are initialized within the domain using
+    RSP. This initial configuration of spheres is then used as a starting
+    point for CRP using Jodrey and Tory's algorithm [1]_.
+
+    In RSP, sphere centers are placed one by one at random, and placement
+    attempts for a sphere are made until the sphere is not overlapping any
+    others. This implementation of the algorithm uses a mesh over the domain
+    to speed up the nearest neighbor search by only searching for a sphere's
+    neighbors within that mesh cell.
+
+    In CRP, each sphere is assigned two diameters, an inner and an outer,
+    which approach each other during the simulation. The inner diameter,
+    defined as the minimum center-to-center distance, is the true diameter of
+    the spheres and defines the pf. At each iteration the worst overlap
+    between spheres based on outer diameter is eliminated by moving the
+    spheres apart along the line joining their centers. Iterations continue
+    until the two diameters converge or until the desired pf is reached.
+
+    References
+    ----------
+    .. [1] W. S. Jodrey and E. M. Tory, "Computer simulation of close random
+       packing of equal spheres", Phys. Rev. A 32 (1985) 2347-2351.
+
+    """
+    # Seed RNG
+    random.seed(seed)
+
+    # Create container with the correct shape based on the supplied region
+    domain = None
+    for cls in _Container.__subclasses__():
+        try:
+            domain = cls.from_region(region, radius)
+        except ValueError:
+            pass
+
+    if not domain:
+        raise ValueError('Could not map region {} to a container: supported '
+                         'container shapes are rectangular prism, cylinder, '
+                         'sphere, and spherical shell.'.format(region))
+
+    # Determine the packing fraction/number of spheres
+    volume = 4/3*pi*radius**3
```

Comment:
Might be nice to have the sphere volume computation as a function as it's used quite a bit.